### ISSUE TITLE
Feature Discussion: Raycast Sensor Performance in mjlab for Parallel RL Training

### ISSUE THREAD

**Issue by kunen23** [mujocolab/mjlab#486]
Thank you very much for maintaining mjlab.
Are there any existing or planned alternatives for faster terrain perception in mjlab? 
It seems that the mjlab + mjwarp raycast sensor (for example height scanner) results in a 2–3× slower overall step time compared to Isaac Lab + Isaac Sim.

**Comment by kevinzakka** [comment]
Hi @kunen23, I haven’t yet had a chance to run a thorough performance benchmark of the raycast sensor in this configuration. If you have a minimal reproducible example or benchmark code you can share, I’d be happy to take a closer look and help diagnose where the slowdown is coming from.

That said, there are a couple of important considerations to keep in mind:
* The Isaac raycast sensor currently only supports static meshes, whereas the mjlab + mjwarp sensor supports dynamic geometry, which comes with additional computational cost.
* There is an upcoming PR to mjwarp that introduces a BVH-based acceleration structure, which should significantly improve raycast performance.

In the meantime, if you can share code or profiling results, I’m very happy to help debug and optimize the current setup.

**Comment by kevinzakka** [comment]
Here's a video comparison using the upcoming BHV for acceleration. I did it on a M1 mac so there's no GPU but the difference is striking:

https://github.com/user-attachments/assets/a8c4e823-f68e-4f8b-853a-f11890d1c44d

**Comment by kunen23** [comment]
Thank you for your quick response @kevinzakka . I write a simple warp raycast sensor for static mesh and it should be faster. It's similar to  RayCastSensor. 

[warp_mesh_raycast_sensor.py](https://github.com/user-attachments/files/24633444/warp_mesh_raycast_sensor.py)

**Comment by kevinzakka** [comment]
@kunen23 Have you benchmarked it?

**Comment by kevinzakka** [comment]
Hi @kunen23, can't merge it just yet but I added the BVH-accelerated raycaster in branch [bvh-raycast](https://github.com/mujocolab/mjlab/tree/bvh-raycast). Note unlike your solution (and the ISAAC one), it is not constrained to static terrain.

**Comment by kunen23** [comment]
Thanks @kevinzakka , I’ll check this branch and benchmark the sensors.
However, I’m currently encountering some issues when training on rough terrain.
In the [bvh-raycast](https://github.com/mujocolab/mjlab/tree/bvh-raycast) branch, should mujoco-warp be based on the [render](https://github.com/google-deepmind/mujoco_warp/tree/render) branch?

https://github.com/user-attachments/assets/30dbbad0-46b2-4999-90c9-b71a77bab8e2

**Comment by kevinzakka** [comment]
Use my fork: https://github.com/kevinzakka/mujoco_warp/tree/bvh-ray

**Comment by Kunen-Peng** [comment]
Thank you for your fork, @kevinzakka.

My setup:

OS: Ubuntu 24.04

CPU: Intel i9-14900KF

GPU: RTX 5070 Ti (16 GB)

I compared the tasks Mjlab-Velocity-Rough-Unitree-Go1 and Mjlab-Velocity-Rough-Unitree-Go1-Warp-Ray.
The only difference between them is that Mjlab-Velocity-Rough-Unitree-Go1-Warp-Ray uses a raycast implementation written by myself.

I ran both tasks with 1024 environments. In the plot, the green line corresponds to Mjlab-Velocity-Rough-Unitree-Go1, and the orange line corresponds to Mjlab-Velocity-Rough-Unitree-Go1-Warp-Ray. It appears that the custom raycast is faster at the beginning, but its performance becomes unstable over time.
Here is the code [bvh-raycast](https://github.com/kunen23/mjlab/tree/bvh-raycast)
![Image](https://github.com/user-attachments/assets/bd7164f2-e896-4435-8b62-1c9e78b5d97d)

**Comment by kevinzakka** [comment]
Hi @kunen23, another thing that will help significantly and that I just merged in head: [sensor caching](9f089d28ef2f83aba46f48add6a27d6bb63a86e4). If the raycast sensor was being accessed multiple times within the same step (e.g. once for the policy, once for the critic, once for the reward), then we'd do a re-computation for all 3 steps. We now cache sensor data within the same step. You should see a non-trivial speedup if you rebase your branch.

**Comment by Kunen-Peng** [comment]
Got it, @kevinzakka. And I forgot to add a termination for rough terrain when the robot goes out of bounds. This should be helpful.
# terrain_out_of_bounds

```python
def terrain_out_of_bounds(
    env: ManagerBasedRlEnv,
    asset_cfg: SceneEntityCfg = _DEFAULT_ASSET_CFG,
    distance_buffer: float = 3.0,
) -> torch.Tensor:
    """Terminate when the actor moves too close to the edge of the terrain.

    If the actor moves too close to the edge of the terrain, the termination is activated.
    The distance to the edge of the terrain is calculated based on the size of the terrain
    and the distance buffer.

    Args:
        env: The environment instance.
        asset_cfg: The configuration of the asset (robot) in the scene.
        distance_buffer: The buffer distance from the terrain edge to trigger termination.

    Returns:
        A tensor of booleans indicating whether each environment should terminate.
    """
    if env.scene.terrain.cfg.terrain_type == "plane":
        return torch.zeros(env.num_envs, dtype=torch.bool, device=env.device)

    elif env.scene.terrain.cfg.terrain_type == "generator":
        # obtain the size of the sub-terrains
        terrain_gen_cfg = env.scene.terrain.cfg.terrain_generator
        grid_width, grid_length = terrain_gen_cfg.size
        n_rows, n_cols = terrain_gen_cfg.num_rows, terrain_gen_cfg.num_cols
        border_width = terrain_gen_cfg.border_width

        # compute the size of the map
        map_width = n_rows * grid_width + 2 * border_width
        map_height = n_cols * grid_length + 2 * border_width

        # extract the used quantities (to enable type-hinting)
        asset: Entity = env.scene[asset_cfg.name]

        # check if the agent is out of bounds
        x_out_of_bounds = (
            torch.abs(asset.data.root_link_pose_w[:, 0])
            > 0.5 * map_width - distance_buffer
        )
        y_out_of_bounds = (
            torch.abs(asset.data.root_link_pose_w[:, 1])
            > 0.5 * map_height - distance_buffer
        )

        return torch.logical_or(x_out_of_bounds, y_out_of_bounds)

    else:
        raise ValueError(
            "Received unsupported terrain type, must be either 'plane' or 'generator'."
        )

**Comment by Kunen-Peng** [comment]
Hi @kevinzakka, are you planning to update the raycast sensor to leverage the new features in MuJoCo Warp https://github.com/google-deepmind/mujoco_warp/pull/1113?

**Comment by kevinzakka** [comment]
Yes! My next big task is updating the raycast sensor and merging the camera sensor!

**Comment by Kunen-Peng** [comment]
Awesome ！! Really looking forward to the raycast sensor update and the camera sensor merge 🚀
Let me know if there’s anything I can help with.

**Comment by energypantry** [comment]
Hi! I can take this on with a fast turnaround. Do you accept USDC/USDT (any EVM chain) for payment? If yes, I can start immediately. Offer page: https://gist.github.com/energypantry/4b187ab7d0629eb793af2c65b42866db

**Comment by kevinzakka** [comment]
Hi @Kunen-Peng, BVH-accelerated ray casting has landed in main. Let me know how it goes!

**Comment by HaoLiuHust** [comment]
Hi, what's the performance compared with IsaacLab？
